contents/c++/rvalue.md

```diff
@@ -0,0 +1,145 @@
+
+Some of the advantages of using Value Semantics is that it ensures that we no memory management issue, as we won't have any [dangling references](https://www.quora.com/What-is-dangling-reference) to objects that may not exist and no [memory leaks](https://www.geeksforgeeks.org/what-is-memory-leak-how-can-we-avoid/).
+
+Value Semantics also ensures [referential transparency](https://en.wikipedia.org/wiki/Referential_transparency), since having our own copy of the value ensures changing the value in 1 object and function will not affect the original value. This is especially useful in a [multi-threaded](https://stackoverflow.com/questions/1313062/what-is-a-multithreaded-application) environment, as this prevents removes the need for synchronization of the values, allowing the program to run faster.
+
+Another seemingly odd benefit of Value Semantics is that it may speedup programs. If we require accessing the value many times, having your own local copy of the value may be faster than having a pointer to the value and dereferencing it each time. This is especially true in C++ as it encourages Value Semantics, with optimization techniques such as [copy elision](http://en.cppreference.com/w/cpp/language/copy_elision) to help passing-by-value be faster.
+
+However, [1 major flaw](https://www.quora.com/What-are-the-drawbacks-of-pass-by-value-result) of Value Semantics and passing-by-value is having poor performance and scability. When we wish to only read the values without modifying it, creating an additional copy just for this purpose is an unnecessary computation, especially if we scale up to larger objects.
+
+### Reference Semantics
+Reference (or pointer) semantics is another mode available for users in C++. C++ allows users to declare [pointers](http://www.cplusplus.com/doc/tutorial/pointers/) and [references](https://www.geeksforgeeks.org/references-in-c/) to the address of the value. As such, we can pass around these pointers to functions and objects, and all of them will refer to and use the same value (address).
+
+One large advantage of using Reference Semantics is that it allows us pass around large objects without having to create an additional copy, thus improving on the performance and scability of the program. It also allows objects to interact with each other as they can have pointers to and modify the same value.
+
+However, Reference Semantics also comes with its own problems. In a multi-threaded environment, having such object pointing to the same value will lead to unnecessary data races and un-deterministic behaviour, and requires extra work in trying to synchronize the values between objects. More details of this can be found in the [Java Concurrency](java/JavaConcurrency.md) and [Java Synchronization](java/JavaSynchronization.md).
```

Comment:
>having such object 

Having a reference or object? And do you mean many references or just a single one?

**indeterminant** behaviour which requires extra work to synchronize values between objects

I'm not sure what `values between objects` mean. Maybe it could be phrased better?